zuko/bayesian.py

```diff
@@ -0,0 +1,391 @@
+                sampled[name + ".weight"] = w_mu + torch.exp(w_logvar).sqrt() * torch.randn_like(
+                    w_mu
+                )
+
+                # Bias - always read from base model unless learn_means=True
+                if module.bias is not None:
+                    if self.learn_means and safe_name + "_bias" in self.bias_means:
+                        b_mu = self.bias_means[safe_name + "_bias"]
+                    else:
+                        b_mu = module.bias
+
+                    # Apply same softclip as BayesianLinear and use sqrt of exp(logvar)
+                    b_logvar = _softclip(self.bias_logvars[safe_name + "_bias"], 11)
+                    sampled[name + ".bias"] = b_mu + torch.exp(b_logvar).sqrt() * torch.randn_like(
+                        b_mu
+                    )
+
+        return sampled
+
+    @contextmanager
+    def sample(self):
+        """Context manager yielding the model with reparametrized forward methods for linear model.
+        This is not cloning the model but just temporarily replacing the forward methods of the
+        selected linear layers to use the local reparameterization trick during training, or
+        sampling once during evaluation."""
+        with self._reparameterization_context():
+            yield self.base
+
+    @contextmanager
+    def _reparameterization_context(self):
+        """Context manager that temporarily replaces linear layer forwards with reparameterized versions."""
+        # Store original forward methods and replace them
+        original_forwards = {}
+
+        for name, module in self.base.named_modules():
+            if isinstance(module, (Linear, MaskedLinear)):
+                # Skip non-selected layers when a subset was provided
+                if self._bayesian_names is not None and name not in self._bayesian_names:
+                    continue
+
+                original_forwards[name] = module.forward
+
+                # Create closure to capture current name and module
+                def make_reparameterized_forward(module_name, original_module):
+                    # Match original forward signature (x) to avoid type errors
+                    if self.training:
+
+                        def reparameterized_forward(x):
+                            return self._apply_local_reparameterization_training(
+                                x, module_name, original_module
+                            )
+
+                    else:
+
+                        def reparameterized_forward(x):
+                            return self._apply_local_reparameterization_eval(
+                                x, module_name, original_module
+                            )
+
+                    return reparameterized_forward
+
+                # Replace the forward method using setattr on instance to avoid mypy complaints
+                module.forward = make_reparameterized_forward(name, module)
```

Comment:
Instead of using a closure, you can use `functools.partial`:

```suggestion
                    if self.training:
                        model.forward = partial(
                            self._apply_local_reparameterization_training,
                            name=name,
                            module=module,
                        )
                    else:
                        model.forward = partial(
                            self._apply_local_reparameterization_eval,
                            name=name,
                            module=module,
                        )
```